### ISSUE TITLE
Rework the packing function hierarchy

### ISSUE THREAD

**Issue by wrtobin** [GEOS-DEV/GEOS#1098]
We've discussed redeveloping the packing functions for a while.

As part of this effort we should:
 - [ ] Modify the `packByIndex` index set argument to allow at least both `ArrayView` and `SortedArrayView` to be used to pack, in order to allow time history collection to not retain an internal copy of the indices being collected.
 - IIRC we also use the internal copy of the packed indices to filter the ghost indices out, so that mechanism needs to be accounted for in the time-hist functionality when removing the internal set copy.
 - Really this is about the higher-level packing functions in e.g. Group, which is virtual.
 - [ ] Refactor the packing functions to use an `array1d` to pack into instead of a raw pointer (#1285). 
 -  [ ] [ Additional cleanup items go here .. ]

**Comment by rrsettgast** [comment]
@wrtobin Is this done?

**Comment by wrtobin** [comment]
No. I think we wanted to also work on the overall call-stack for the packing functions starting from the group/wrapper-level to just refactor the entire thing at once, but haven't taken the time to get into it yet. I can follow up once the HIP branch is merged in.
